torch/_inductor/comms.py

```diff
@@ -17,61 +20,124 @@
-    new_order_reversed.extend(tuple_sorted(cur_comms))
-    return new_order_reversed[::-1]
+    return raise_comms_and_sink_waits(snodes, raise_comms=True, sink_waits=False)
+
+
+def raise_comms_and_sink_waits(
+    snodes: List[BaseSchedulerNode],
+    *args,
+    raise_comms: bool = True,
+    sink_waits: bool = True,
+) -> List[BaseSchedulerNode]:
+    # We assign each node a tuple of scores (score_0, score_1, score_2),
+    # decreasing in importance, with a lower value indicating a higher ranking:
+    #
+    # - score_0: the lowest comm_idx among the comm nodes that the node blocks.
+    # If a node doesn't block any comm nodes, its score_0 is set to
+    # sys.maxsize. This score ensures that comm nodes get scheduled as early as
+    # possible.
+    # - score_1: 1 if the node is a wait node, 0 otherwise. This score ensures
+    # that wait nodes are deferred as late as possible.
+    # - score_2: the index of the node in the original topological order. This
+    # score provides stability in case of ties.
+    #
+    # When only raise_comms is True, only score_0 and score_2 are considered.
+    # When only sink_waits is True, only score_1 and score_2 are considered.
+    # When neither is True, the original order is yielded.
+    name_to_snode = {}
+    scores_0, scores_1, scores_2 = {}, {}, {}
+    for idx, snode in enumerate(snodes):
+        for name in snode.get_names():
+            name_to_snode[name] = snode
+            scores_0[name] = sys.maxsize
+            scores_1[name] = 0
+            scores_2[name] = idx
+
+    comm_idx = 0
+    for snode in snodes:
+        if raise_comms and is_collective(snode.node):
+            scores_0[snode.get_name()] = comm_idx
+            for anc in snode.ancestors:
+                scores_0[anc] = min(scores_0[anc], comm_idx)
+            comm_idx += 1
+        elif sink_waits and is_wait(snode.node):
+            scores_1[snode.get_name()] = 1
+
+    class Runnable:
+        def __init__(self, snode):
+            self.snode = snode
+            name = next(iter(snode.get_names()))
+            self.score = (
+                scores_0[name],
+                scores_1[name],
+                scores_2[name],
+            )
+
+        def __lt__(self, other):
+            return self.score < other.score
+
+    # A mutating node's unmet_dependencies doesn't cover the dependencies
+    # caused by the mutation. Instead, they are described by associated
+    # MutationOutput node. Thus, to safely schedule a mutating node, we have to
+    # add the unmet_dependencies of the associated MutationOutput nodes to the
+    # mutating node.
+    # TODO(yifu): this is needed due to a mutation handling bug in the
+    # scheduler. It should be fixed by https://github.com/pytorch/pytorch/pull/128893.
+    # We can remove this logic once the fix is landed.
+    unmet_deps: Dict[BaseSchedulerNode, Set[str]] = {}
+    for snode in snodes:
+        if isinstance(snode.node, ir.MutationOutput):
```

Comment:
Do we still need this change after @peterbell10's fix?

Reply:
No we can remove it after the fixes land. @yf225 and I decided to land this version first since we both have changes that depend on this PR (also this PR fixes the tests and enables them in CI).